internal/parser/parser.go

```diff
@@ -207,25 +207,30 @@ func findPort(svc *corev1.Service, wantPort kongstate.PortDef) (*corev1.ServiceP
 
 func getUpstreams(
 	log logrus.FieldLogger, s store.Storer, serviceMap map[string]kongstate.Service) []kongstate.Upstream {
-	var upstreams []kongstate.Upstream
+	upstreamDedup := make(map[string]struct{})
```

Comment:
Technically when there's a very large number of upstreams we would get some thrash from the go runtime making new arrays and copying our deduplication map to the larger arrays. The performance hit here would probably would be fairly minor or maybe not even noticeable, but since we can just as easily avoid it at the cost of _maybe_ a few extra slots on the initial array that don't end up getting used (duplicates are fairly uncommon, as per the nature of this issue which we consider a bit of an edge case) we should:
```suggestion
	upstreamDedup := make(map[string]struct{}, len(serviceMap))
```